File: app/src/main/java/cz/ackee/sample/detail/DetailPresenter.java
package cz.ackee.sample.detail;

import java.util.List;

import cz.ackee.sample.interactor.ApiInteractorImpl;
import cz.ackee.sample.login.ILoginView;
import cz.ackee.sample.model.SampleItem;

/**
 * TODO add class description
 * Created by David Bilik[david.bilik@ackee.cz] on {05/08/16}
 **/
public class DetailPresenter {
    public static final String TAG = DetailPresenter.class.getName();
    private final ApiInteractorImpl apiInteractor;
    private IDetailView view;
    private List<SampleItem> items;

    public DetailPresenter() {
        this.apiInteractor = new ApiInteractorImpl();
    }

    public void onViewAttached(IDetailView view) {
        this.view = view;
        if(items != null) {
            view.showData(items);
        }
    }

    public void onViewDetached() {
        this.view = null;
    }

    public void refresh() {
        apiInteractor
                .getData()
                .subscribe(this::onDataLoaded, this::onErrorHappened);
    }

    private void onErrorHappened(Throwable throwable) {
        throwable.printStackTrace();
    }

    private void onDataLoaded(List<SampleItem> sampleItems) {
        this.items = sampleItems;
        if (view != null) {
            view.showData(sampleItems);
        }
    }
}

File: app/src/main/java/cz/ackee/sample/detail/DetailPresenter.kt
package cz.ackee.sample.detail

import cz.ackee.sample.App
import cz.ackee.sample.model.SampleItem
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers

/**
 * Presenter for detail screen
 */
class DetailPresenter {

    private val apiInteractor = App.diContainer.apiInteractor
    private var view: IDetailView? = null
    private var items: List<SampleItem>? = null

    fun onViewAttached(view: IDetailView) {
        this.view = view
        if (items != null) {
            view.showData(items!!)
        }
    }

    fun onViewDetached() {
        this.view = null
    }

    fun refresh() {
        apiInteractor.getData()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ this.onDataLoaded(it) }, { this.onErrorHappened(it) })
    }

    private fun onErrorHappened(throwable: Throwable) {
        throwable.printStackTrace()
    }

    private fun onDataLoaded(sampleItems: List<SampleItem>) {
        this.items = sampleItems
        view?.showData(sampleItems)
    }

    fun invalidateAccessToken() {
        with(App.diContainer.oauthStore) { saveOauthCredentials("bla", refreshToken ?: "") }
    }

    fun invalidateRefreshToken() {
        with(App.diContainer.oauthStore) { saveOauthCredentials(accessToken ?: "", "bla") }
    }

    fun logout() {
        apiInteractor.logout()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { App.diContainer.logouter.logout() }
    }
}
